module-system/sov-state/src/storage.rs

```diff
@@ -29,7 +29,13 @@ impl StorageKey {
         self.key.clone()
     }
 
-    pub fn as_cache_key(self) -> CacheKey {
+    pub fn as_cache_key(&self) -> &CacheKey {
+        // Safety: they are currently equivalent
+        // TODO https://github.com/Sovereign-Labs/sovereign-sdk/issues/643
+        unsafe { core::mem::transmute(self) }
```

Comment:
Since we don't use `#[repr(transparent)]`, I think rustc is allowed to use whatever layout it wants for storage keys; so while they're likely the same as cache keys in practice we can't guarantee that.

Reply:
Since there all just wrapper around arc, wouldn't cloning arc just work?

Reply:
@preston-evans98 I think it's ok for this case since the type declaration is the same. The transparent would be required when we convert from struct to enum, for example. In that case, the compiler would, maybe, use a different layout.

@citizen-stig strictly speaking, cloning the arc into an owned cache key would be just fine as internally it would just increment the reference count to the vec. I'll dive into more details of what I had in mind, but first, a preliminary snippet:

```rust
pub fn as_cache_key(&self) -> &CacheKey {
    &CacheKey { key: self.key }
}
```

The snippet above doesn't work because `CacheKey` is created inside the function, even tho it contains just a pointer that will be valid for the lifetime of `StorageKey`.

The idea of having references to the storage key instead of owned storage key has two main pillars:

- The assumption of storage key being a pointer is an implementation detail of the struct; otherwise, it would implement `Copy`
- According to the snippet below, we need a reference to the Arc to satisfy the API constraints
```rust
use std::{borrow::Borrow, marker::PhantomData, sync::Arc};

#[derive(Default)]
pub struct Map<K, V> {
    _data: PhantomData<(K, V)>,
}

impl<K, V> Map<K, V> {
    pub fn get<Q>(&self, _k: &Q) -> Option<V>
    where
        Q: ?Sized,
        K: Borrow<Q>,
    {
        todo!()
    }
}

Map::<Vec<u8>, u8>::default().get(&[0][..]); // works fine
Map::<Arc<Vec<u8>>, u8>::default().get(&Arc::new(vec![0u8])); // works too since arc borrows
                                                              // whatever T borrows
Map::<Arc<Vec<u8>>, u8>::default().get(Arc::new(vec![0u8])); // doesn't work since we expect a
                                                             // reference to a borrowable type
```

The idea of making `StorageKey` always a reference challenges the assumption passed to the user that cloning this key is fine (in reality it is, but as mentioned above, it is an implementation detail of the struct).

Also, the idea of making `StorageKey` always a reference is more ergonomic to work with map keys that will, [by default](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get), expect references of generics.

Wdyt of this approach?

---

There is also a rather questionable point with the PR. The previous API was using `as_cache_key` to convert owned -> owned. Even tho it is a sort of free conversion (pointer to pointer), it is still owned to owned and should be `to_*`.

Reply:
Finally, this solution is meant to be a cheap one until we finally decide what to do with #643 . If we unify the types, we could even implement copy for the struct and treat it as a pointer

Reply:
> I think it's ok for this case since the type declaration is the same. The transparent would be required when we convert from struct to enum, for example. In that case, the compiler would, maybe, use a different layout.

Logically I agree, espeically that there's only single inner attribute.

But I still think that this reasoning might be not 100% sound, because instead of relying on contract with compiler, we rely on our understanding of compiler. 
Maybe let's just add `repr(transparent)` to be on the safe side?

Reply:
Sounds good! I added the transparent directive and a test-only assertion that will pass only if the two structs are exactly equal

https://github.com/Sovereign-Labs/sovereign-sdk/pull/644/commits/7aae44d65720e122bfec5a55f1e3845f65412b68

Reply:
Looks great, thanks!

Reply:
To recap, two options were proposed here:

1. Clone the `Arc`.
2.  `std::mem::transmute`.

I appreciate the the safeguards we added here, and I'm actually convinced it's safe (MIRI agrees), but we're still relying on the internals of the two structs as an implementation detail during the transmute. The first option seems simpler to understand, 100% safe, and quite intuitive. It relies on an implementation detail, which is unfortunate, but the same can be said about this `unsafe`.

I'd be for dropping this `unsafe` and clone the `Arc` instead, which is a safer stopgap until #643 lands.